contracts/lotterybook/merkletree/merkletree.go

```diff
@@ -0,0 +1,377 @@
+func (s EntryByBias) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// EntryByLevel implements the sort interface to allow sorting a list of entries
+// by their position in the tree in descending order.
+type EntryByLevel []*Entry
+
+func (s EntryByLevel) Len() int           { return len(s) }
+func (s EntryByLevel) Less(i, j int) bool { return s[i].level > s[j].level }
+func (s EntryByLevel) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// Node represents a node in merkle tree.
+type Node struct {
+	Nodehash common.Hash // The hash of node.
+	Parent   *Node       // The parent of this node, nil if it's root node.
+	Left     *Node       // The left child of this node
+	Right    *Node       // The right child of this node
+	Level    uint64      // The level of node in this tree
+	Value    *Entry      // The referenced entry by this node, nil if it's not leaf.
+}
+
+// Hash returns the hash of this tree node.
+func (node *Node) Hash() common.Hash {
+	// Short circuit if nodehash is already cached.
+	if node.Nodehash != (common.Hash{}) {
+		return node.Nodehash
+	}
+	// If it's a leaf node, derive the hash by the entry content.
+	if node.Value != nil {
+		node.Nodehash = node.Value.Hash()
+		return node.Nodehash
+	}
+	// It's a branch node, derive the hash via two children.
+	leaf, right := node.Left.Hash(), node.Right.Hash() // Both children should never be nil.
+	if bytes.Compare(leaf.Bytes(), right.Bytes()) < 0 {
+		node.Nodehash = crypto.Keccak256Hash(append(leaf.Bytes(), right.Bytes()...))
+	} else {
+		node.Nodehash = crypto.Keccak256Hash(append(right.Bytes(), leaf.Bytes()...))
+	}
+	return node.Nodehash
+}
+
+// String returns the string format of node.
+func (node *Node) String() string {
+	if node.Value != nil {
+		return fmt.Sprintf("E(%x:%d)", node.Value.Value, node.Value.level)
+	}
+	return fmt.Sprintf("N(%x) => L.(%s) R.(%s)", node.Hash(), node.Left.String(), node.Right.String())
+}
+
+type MerkleTree struct {
+	Roothash common.Hash // The hash of root node, maybe null if we never calculate it.
+	Root     *Node       // The root node of merkle tree.
+	Leaves   []*Node     // Batch of leaves node included in the tree.
+}
+
+// NewMerkleTree constructs a merkle tree with given entries.
+func NewMerkleTree(entries []*Entry) (*MerkleTree, error) {
+	if len(entries) == 0 {
+		return nil, ErrEmptyEntryList
+	}
+	// Assign an unique salt for each entry. The hash
+	// of entry is calculated by keccak256(value, salt)
+	// so we can preserve the privacy of given value.
+	for _, entry := range entries {
+		entry.salt = rand.Uint64()
+	}
+	// Verify the validity of the given entries.
+	var sum, totalWeight uint64
+	for _, entry := range entries {
+		if entry.Weight == 0 {
+			return nil, ErrInvalidWeight
+		}
+		sum += entry.Weight
+	}
+	for _, entry := range entries {
+		l := math.Log2(float64(sum) / float64(entry.Weight))
+		c := math.Ceil(l)
+		entry.bias = l - c + 1
+		if int(c) > maxLevel {
+			return nil, ErrInvalidWeight
```

Comment:
> It is better to either give the minimum value slot (1/1024) or zero allowance to such items

I personally prefer to reject the construction in the first place. E.g. If we assign a zero allowance it may break the code that we hold the assumption if the receiver is included but actually it's not.

Reply:
I agree that it is generally a good practice to not allow a situation to happen that could be prevented earlier. But I think the assumptions are different here. The way we are going to use the tree creation is that we have a number of payment candidates with weights calculated based on previously collected data. The number of nodes can be large or small, the weights can also be either significant or very small. Sure, we have to filter out very small weights somewhere but the thing is that there is no definite threshold to say that this payment route will be useful or not. Even a very small weight can have a little bit of utility because we can at least try the paid service of one more server. On the other hand, filtering them before the tree creation is actually harder because I have to know how the tree works and calculate the sum of weights and check whether some items are below 1/1024 of that. I also can't know whether maybe the item that is slightly below this threshold could still get a slot. It is a lot better to do the weight filtering inside the tree creation (which happens automatically in my version of the code). Then the caller does not need to know about the internal details of the lottery tree, it does not need to be prepared for an error, it just needs to be prepared that what it gets might not be exactly the same as what it has requested (but that is true anyways).

Reply:
I wan to confirm that in your version, if the amount is too small, there is no guarantee that it will eventually be assigned a non-zero weight right?

Reply:
Yes